### PR TITLE
Preparation for Light client RPC

### PR DESCRIPTION
**Pull request by rphmeier**
This PR gets us most of the way to RPCs for the current light client, and ports the existing full RPCs over to more generic abstractions instead of `MiningBlockChainClient` and `MinerService`.

@tomusdrw This touches a lot of code in the signer implementation (particularly to move it to futures). The tests pass, but let me know if you see anything suspicious. There will also undoubtedly be conflicts with #4477.

Key points:
- Skeleton of `eth` API
- Generic `Dispatcher` with `FullDispatcher` implementation
- `OnDemand` improvements

Next up:
  - Light client transaction queue
  - OnDemand service request batching
  - `LightDispatcher` (fetch nonce, historic gas price via OnDemand)

Part of #2235

### CODE REVIEW

File: ethcore/light/src/client/mod.rs [hunk 1]
```diff
@@ -109,12 +107,12 @@ impl Client {
...
 	pub fn ready_transactions(&self) -> Vec<PendingTransaction> {
-		let best_num = self.chain.best_block().number;
+		let best = self.chain.best_header();
 		self.tx_pool.lock()
 			.values()
 			.filter(|t| match t.condition {
-				Some(TransactionCondition::Number(ref x)) => x <= &best_num,
-				Some(TransactionCondition::Timestamp(ref x)) => *x <= time::get_time().sec as u64,
+				Some(TransactionCondition::Number(x)) => x <= best.number(),
+				Some(TransactionCondition::Timestamp(x)) => x <= best.timestamp(),
```
Comment (tomusdrw) [inline ethcore/light/src/client/mod.rs]: I think this should be a separate PR to have some discussion about it. 
IMHO it's not the right way to do it, we may start propagating transaction way too late.

File: rpc/src/v1/helpers/dispatch.rs [hunk 2]
```diff
@@ -38,20 +41,142 @@ use v1::types::{
...
+		future::done(inner()).boxed()
+	}
+
+	fn dispatch_transaction(&self, signed_transaction: PendingTransaction) -> Result<H256, Error> {
+		let hash = signed_transaction.transaction.hash();
+
+		take_weak!(self.miner).import_own_transaction(&*take_weak!(self.client), signed_transaction)
+			.map_err(errors::from_transaction_error)
+			.map(|_| hash)
+		}
```
Comment (tomusdrw) [inline rpc/src/v1/helpers/dispatch.rs]: Incorrect indentation

File: rpc/src/v1/impls/eth.rs [hunk 3]
```diff
@@ -356,113 +356,144 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM> Eth for EthClient<C, SN, S, M, EM> where
...
+					match client.balance(&address, id.into()) {
+						Some(balance) => Ok(balance.into()),
+						None => Err(errors::state_pruned()),
+					}
 				}
 			}
-		}
+		};
+
+		future::done(inner()).boxed()
```
Comment (tomusdrw) [inline rpc/src/v1/impls/eth.rs]: Maybe it's worth turning this pattern into macro? It will be greatly simplified when `Carrier` is stabilized anyway.

File: rpc/src/v1/impls/personal.rs [hunk 4]
```diff
@@ -106,28 +95,41 @@ impl<C, M> Personal for PersonalClient<C, M> where
 	}
 
 	fn send_transaction(&self, meta: Metadata, request: TransactionRequest, password: String) -> BoxFuture<RpcH256, Error> {
-		let sign_and_send = move || {
-			let client = take_weak!(self.client);
-			let miner = take_weak!(self.miner);
+		let setup = || {
+			let dispatcher = self.dispatcher.clone();
 			let accounts = take_weak!(self.accounts);
```
Comment (tomusdrw) [inline rpc/src/v1/impls/personal.rs]: Perhaps it's worth to introduce `take_weak` alike macro for futures to avoid such `setup` steps.

File: rpc/src/v1/impls/signing.rs [hunk 5]
```diff
@@ -50,105 +49,103 @@ enum DispatchResult {
...
-		let default_account = match default_account {
-			DefaultAccount::Provided(acc) => acc,
-			DefaultAccount::ForDapp(dapp) => take_weak!(self.accounts).default_address(dapp).ok().unwrap_or_default(),
+			Ok((self.dispatcher.clone(), accounts, default_account))
 		};
-		let payload = dispatch::from_rpc(payload, default_account, &*client, &*miner);
-		self.add_to_queue(payload)
+
+		let weak_signer = self.signer.clone();
+		future::done(setup())
```
Comment (tomusdrw) [inline rpc/src/v1/impls/signing.rs]: Identation here looks bad.
Could be simplified with `take_weak` for futures

```rust
let accounts = take_weakf!(self.accounts);
let signer = take_weakf!(self.signer);
let default_account = match default_account {
  DefaultAccount::Provided(acc) => acc,
  DefaultAccount::ForDapp(dapp) => accounts.default_address(dapp).ok().unwrap_or_default(),
};

dispatch::from_rpc(payload, default_account, &self.dispatcher)
  .and_then(|payload| {
     ...
   })
   .boxed()
```

File: rpc/src/v1/impls/signing.rs [hunk 6]
```diff
@@ -170,67 +167,78 @@ impl<C: 'static, M: 'static> ParitySigning for SigningQueueClient<C, M> where
...
-				e => ready.complete(Err(errors::internal("Unexpected result.", e))),
-			}
-		});
-
-		p.then(|result| futures::done(result.expect("Ready is never dropped nor canceled."))).boxed()
+
+		// when dispatch is complete
+		res.then(move |res| {
+			// register callback via the oneshot sender.
+			handle_dispatch(res, move |response| {
```
Comment (tomusdrw) [inline rpc/src/v1/impls/signing.rs]: Most probably `handle_dispatch` could now return Future instead of accepting a callback (but it's a separate issue)

File: rpc/src/v1/impls/signing.rs [hunk 7]
```diff
@@ -170,67 +167,78 @@ impl<C: 'static, M: 'static> ParitySigning for SigningQueueClient<C, M> where
...
+		res.then(move |res| {
+			handle_dispatch(res, move |response| {
+				match response {
+					Ok(RpcConfirmationResponse::Signature(sig)) => ready.complete(Ok(sig)),
+					Err(e) => ready.complete(Err(e)),
+					e => ready.complete(Err(errors::internal("Unexpected result.", e))),
+				}
+			});
+
+			p.then(|result| futures::done(result.expect("Ready is never dropped nor canceled."))).boxed()
```
Comment (tomusdrw) [inline rpc/src/v1/impls/signing.rs]: `.boxed()` unecessary

File: rpc/src/v1/impls/signing.rs [hunk 8]
```diff
@@ -170,67 +167,78 @@ impl<C: 'static, M: 'static> ParitySigning for SigningQueueClient<C, M> where
...
+			handle_dispatch(res, move |response| {
+				match response {
+					Ok(RpcConfirmationResponse::Decrypt(data)) => ready.complete(Ok(data)),
+					Err(e) => ready.complete(Err(e)),
+					e => ready.complete(Err(errors::internal("Unexpected result.", e))),
+				}
+			});
+
+			// and wait for that to resolve.
+			p.then(|result| futures::done(result.expect("Ready is never dropped nor canceled."))).boxed()
```
Comment (tomusdrw) [inline rpc/src/v1/impls/signing.rs]: Unecessary `.boxed()`

File: rpc/src/v1/impls/signing_unsafe.rs [hunk 9]
```diff
@@ -38,106 +37,101 @@ use v1::types::{
...
 	}
 
-	fn handle(&self, payload: RpcConfirmationPayload, account: DefaultAccount) -> Result<RpcConfirmationResponse, Error> {
-		let client = take_weak!(self.client);
-		let miner = take_weak!(self.miner);
-		let accounts = take_weak!(self.accounts);
+	fn handle(&self, payload: RpcConfirmationPayload, account: DefaultAccount) -> BoxFuture<RpcConfirmationResponse, Error> {
+		let setup = move || {
+			let accounts = take_weak!(self.accounts);
+			let default_account = account;
```
Comment (tomusdrw) [inline rpc/src/v1/impls/signing_unsafe.rs]: Why is this line needed? Similar to `dispatch` could be greatly simplified with `take_weak` for futures

File: rpc/src/v1/impls/signing.rs [hunk 10]
```diff
@@ -50,105 +49,103 @@ enum DispatchResult {
...
-					on_response(result.unwrap_or_else(|| Err(errors::request_rejected())))
-				})
-			},
-			Err(e) => on_response(Err(e)),
-		}
-	}
+	fn dispatch(&self, payload: RpcConfirmationPayload, default_account: DefaultAccount) -> BoxFuture<DispatchResult, Error> {
+		let setup = move || {
+			let accounts = take_weak!(self.accounts);
+			let default_account = default_account;
```
Comment (tomusdrw) [inline rpc/src/v1/impls/signing.rs]: Superflous line?

File: rpc/src/v1/tests/mocked/signing.rs [hunk 11]
```diff
@@ -91,9 +95,17 @@ fn should_add_sign_to_queue() {
 
 	// then
 	let promise = tester.io.handle_request(&request);
-	assert_eq!(tester.signer.requests().len(), 1);
-	// respond
-	tester.signer.request_confirmed(1.into(), Ok(ConfirmationResponse::Signature(0.into())));
+
+	// the future must be polled at least once before request is queued.
+	let signer = tester.signer.clone();
+	::std::thread::spawn(move || loop {
```
Comment (tomusdrw) [inline rpc/src/v1/tests/mocked/signing.rs]: We need to figure out how to do it better :/ Will think about it :)

File: rpc/src/v1/impls/signing.rs (hunk 10 above)
Comment (rphmeier) [inline rpc/src/v1/impls/signing.rs]: rustc can't deduce the move of `default_account` without it.

File: rpc/src/v1/impls/signing_unsafe.rs (hunk 9 above)
Comment (rphmeier) [inline rpc/src/v1/impls/signing_unsafe.rs]: forces rustc to recognize the fact that `account` has been moved into the closure, otherwise we'd have to `clone()` it.

File: ethcore/light/src/client/mod.rs (hunk 1 above)
Comment (rphmeier) [inline ethcore/light/src/client/mod.rs]: This is just the change discussed here: https://github.com/ethcore/parity/pull/4419#discussion_r99390569

What's wrong with it?

File: rpc/src/v1/impls/eth.rs (hunk 3 above)
Comment (rphmeier) [inline rpc/src/v1/impls/eth.rs]: Yes, might be worth doing. I don't like to hide the allocation but I suppose it's evident in the return type.

File: rpc/src/v1/impls/personal.rs (hunk 4 above)
Comment (rphmeier) [inline rpc/src/v1/impls/personal.rs]: No, the `setup` phase explicitly allows us to avoid multi-level `BoxFuture`s where possible. If we used a `take_weak`-like macro in an `and_then` combinator, we'd be restricted to having the `and_then` closure return `BoxFuture`.

File: ethcore/light/src/client/mod.rs (hunk 1 above)
Comment (tomusdrw) [inline ethcore/light/src/client/mod.rs]: Pending transactions might be first propagated a little bit too late (we need to first import the block with given timestamp and only after that the propagation starts).
It's probably OK, but we need to make sure to indicate this in docs/ui so that people understand the consequences.

Comment (rphmeier) [inline ethcore/light/src/client/mod.rs]: Yes, but it protects against issues with local time being out-of-sync with "chain time". Documentation is definitely key.

File: rpc/src/v1/impls/signing.rs (hunk 6 above)
Comment (rphmeier) [inline rpc/src/v1/impls/signing.rs]: right, that'll involve a bit more digging into the internals of the signer queue (to get rid of the old async callback stuff) but should be done eventually.

Comment (tomusdrw) [inline rpc/src/v1/impls/signing.rs]: Yup, probably it's best to log it (https://github.com/ethcore/parity/issues/4490) and leave as a separate PR.

File: rpc/src/v1/impls/eth.rs (hunk 3 above)
Comment (rphmeier) [inline rpc/src/v1/impls/eth.rs]: @tomusdrw Can we leave as-is for now? The PR which finishes up the `eth` + `eth_filter` implementation can clean things up. For now, I want to see what patterns actually emerge and need visual optimizing.

**Comment by rphmeier** [comment]
@tomusdrw Grumbles (except for the test stuff) should be addressed now.

File: rpc/src/v1/tests/mocked/signing.rs (hunk 11 above)
Comment (rphmeier) [inline rpc/src/v1/tests/mocked/signing.rs]: Can we log that as a separate issue, at least?

Comment (tomusdrw) [inline rpc/src/v1/tests/mocked/signing.rs]: Sure.

File: rpc/src/v1/impls/eth.rs [hunk 12]
```diff
@@ -356,113 +356,144 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM> Eth for EthClient<C, SN, S, M, EM> where
...
 	}
 
-	fn block_transaction_count_by_hash(&self, hash: RpcH256) -> Result<Option<RpcU256>, Error> {
-		Ok(
-			take_weak!(self.client).block(BlockId::Hash(hash.into()))
-				.map(|block| block.transactions_count().into())
-		)
+	fn block_transaction_count_by_hash(&self, hash: RpcH256) -> BoxFuture<Option<RpcU256>, Error> {
+		let inner = || {
```
Comment (gavofyork) [inline rpc/src/v1/impls/eth.rs]: what's the point of this `inner` pattern? doesn't it just add an extra 3 lines of useless code into every function?

File: rpc/src/v1/impls/eth.rs (hunk 3 above)
Comment (gavofyork) [inline rpc/src/v1/impls/eth.rs]: i'm really not very keen on introducing such a swathe of boilerplate.

**Comment by gavofyork** [comment]
not very happy about the additional of boilerplate. is there nothing that can be done?

**Comment by rphmeier** [comment]
It serves a purpose: convert result-using code into future-using code. Two options:
  - use the just-written `take_weakf!` and pay more interleaved "boilerplate" cost with stuff like `return future::err(e).boxed()` instead of `return Err(e)`
  - write a macro to wrap the function body and compile down to the `let inner = || {...}` pattern used explicitly here. Still introduces the visual noise + rightwards drift of the status quo.

**Comment by tomusdrw** [comment]
There is no additional cost with `teak_weakf!`, in both examples we return one `.boxed()` future. There is an issue to get rid of those `Weak` references: #3958  that would also reduce the boilerplate.

**Comment by rphmeier** [comment]
I understand there's no more runtime cost, I mean you still pay a boilerplate cost in places, just more spread out.

**Comment by rphmeier** [comment]
Ok, no more `let inner` pattern via `take_weakf` and `try_bf` macros.

**Comment by tomusdrw** [comment]
Some conflicts to resolve.

**Comment by rphmeier** [comment]
(resolved locally, but can't build after #4486)
